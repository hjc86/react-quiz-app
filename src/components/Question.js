import React, { Component } from 'react'
import ReactDOM from "react-dom"
import Answer from './Answer'
import '../App.css';
import Score from './Score';
import styles from './Question.module.css';

class Question extends Component {
    constructor(props){
        super(props)
        this.state={
            randomisedAnswers:null,
            correctAnswer: null,
            wrongAnswers: null,
            backgroundStyle: "none",
            clicked: false,
            userAnswers : [],
            questionDecoded: ''
        }
    }

    createAnswerArray= ()=> {  
        let questionData= this.props.questionData 
        let answerArray= questionData.incorrect_answers.concat([questionData.correct_answer])
        let arr = this.getRandomArray(answerArray)
        this.setState(
            {
                randomisedAnswers: arr,
                correctAnswer: questionData.correct_answer,
                wrongAnswers: questionData.incorrect_answers,
            }
        )
    }

    getRandomArray= (array) => {
        const questionAnswerIndicies = []
        const randomisedAnswers = []
        while(questionAnswerIndicies.length<array.length){
           let randomIndex = Math.floor(Math.random() * Math.floor(array.length));
            if(!questionAnswerIndicies.includes(randomIndex)){
                questionAnswerIndicies.push(randomIndex)
                randomisedAnswers.push(array[randomIndex])
            };
        }
         
        let randomisedAnswersDecoded= randomisedAnswers.map((answer)=> {
            let txt = document.createElement("textarea")
            txt.innerHTML = answer
            return txt.value
        })
            
        return randomisedAnswersDecoded
    }

    hasBeenClicked = (answerData)=>{
        this.setState({
            clicked : true,
            selectedAnswer: answerData.selectedAnswer,
            backgroundStyle: answerData.backgroundStyle
        })

        if(answerData.addPoint){
            this.props.latestPoint(1)
        }
        else{
            this.props.latestPoint(0)
        }

        this.props.answeredQuestions(1)
    }

    decodeHtml(html) {
        var txt = document.createElement("textarea");
        txt.innerHTML = html;
        this.setState({
            questionDecoded: txt.value
        })
    }

    componentDidMount(){
        this.createAnswerArray()
        this.decodeHtml(this.props.questionData.question)
    }


    render() {
                return (
                    
                    <div className={styles.questionGroup}>

                    {this.state.randomisedAnswers === null ? "waiting for quiz to be generated" :
                        <React.Fragment>
        
                            <div className={styles.questionTitle}>
        

                                <div className={styles.questionBox}>
                                            
                                    {this.state.questionDecoded}

                                </div>
                            </div>
                            {this.state.randomisedAnswers.map((answer)=> (
                            <Answer {...this.state} answer={answer} clicked={this.hasBeenClicked} className={this.state.clicked ? "inactive": ""} />
                            ))}
                        </React.Fragment>
                    }
                       </div>
                )
            }

}
export default Question