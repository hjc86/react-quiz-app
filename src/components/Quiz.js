import styles from './Quiz.module.css';
import React from 'react';
import Score from'./Score'
import Question from './Question';


class Quiz extends React.Component {

  constructor(props){
    super(props)
  
    this.state ={
      playerScore: 0,
      questionsAnswered: 0,
      quizInstance: null
    }
  }

  updateScore= async (currentScore)=>{
    await this.setState({
      playerScore: this.state.playerScore+currentScore
    }) 

    let playerObj={
      score: this.state.playerScore,
      playerName: this.props.currentPlayer
    }

    this.props.score(playerObj)
    console.log("quiz instance player score", this.state.playerScore)
  }
  
  updateAnsweredQuestions= async (currentAnswered)=>{
    await this.setState({
      questionsAnswered: this.state.questionsAnswered+currentAnswered
    })

    this.props.answeredQuestions(this.state.questionsAnswered) 
  }

  refreshPage = () => {
    window.location.reload(false);
  }

  getQuizApi = async (userData) => {

    const categoryNumber = userData.category;
    const difficulty = userData.difficulty;


    // header('content-type:text/html;charset=utf-8');
    const apiCall = await fetch(`https://opentdb.com/api.php?amount=5&category=${categoryNumber}&difficulty=${difficulty}&type=multiple&dencode=base64`)
    
    const data = await apiCall.json();

    this.setState({
          quizInstance: data.results
    })
  } 


  componentDidMount =()=>{
    
    this.getQuizApi(this.props.playerInfo)
  
  }
  

  render() {
    let categoryName="PLACEHOLDER"

    return (
      this.state.quizInstance === null ? 
            <span className={styles.Bee}> We're BEEZZY loading {this.props.currentPlayer}'s quiz </span>  :

            <div className={styles.quizComponent}>
              
              <div className={styles.quizDeclaration}>
                {this.props.currentPlayer} you are being WHUIZZED on {this.props.playerInfo.categoryName}!

              </div>
              
              <div className="quizQuestionsMapped">
                {this.state.quizInstance.map((questionItem) => 
                  <Question 
                    questionData={questionItem} 
                    latestPoint={this.updateScore} 
                    answeredQuestions={this.updateAnsweredQuestions}/>)
                  }
              </div>
              
        
            
            </div>
          
    
        // <div className={styles.quizDiv}>
        // <h3 className={styles.question}>Question : {this.props.questions}</h3>

        // {/* // display answers - find a way to randomise and space the answers out in button form */}
        // <h3 className={styles.question}>Answer : {this.props.answers} {this.props.incorrect}</h3>
        // </div>
    )
  }
}


export default Quiz;
