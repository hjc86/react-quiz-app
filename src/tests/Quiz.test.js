import React from 'react';
import ReactDOM from 'react-dom'
import { shallow, mount } from 'enzyme';
import Quiz from '../components/Quiz';
import renderer from 'react-test-renderer';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Question from '../components/Question';

let testData ={ 
    "results": [
    {
      "category": "Animals",
      "type": "multiple",
      "difficulty": "easy",
      "question": "Hippocampus is the Latin name for which marine creature?",
      "correct_answer": "Seahorse",
      "incorrect_answers": [
        "Dolphin",
        "Whale",
        "Octopus"
      ]
    },
    {
      "category": "Science: Computers",
      "type": "multiple",
      "difficulty": "easy",
      "question": "If you were to code software in this language you&#039;d only be able to type 0&#039;s and 1&#039;s.",
      "correct_answer": "Binary",
      "incorrect_answers": [
        "JavaScript",
        "C++",
        "Python"
      ]
    },
    {
      "category": "General Knowledge",
      "type": "multiple",
      "difficulty": "easy",
      "question": "Which one of the following rhythm games was made by Harmonix?",
      "correct_answer": "Rock Band",
      "incorrect_answers": [
        "Meat Beat Mania",
        "Guitar Hero Live",
        "Dance Dance Revolution"
      ]
    },
    {
      "category": "Entertainment: Video Games",
      "type": "multiple",
      "difficulty": "easy",
      "question": "Which of these is NOT a playable character in the 2016 video game Overwatch?",
      "correct_answer": "Invoker",
      "incorrect_answers": [
        "Mercy",
        "Winston",
        "Zenyatta"
      ]
    },
    {
      "category": "General Knowledge",
      "type": "multiple",
      "difficulty": "easy",
      "question": "The Flag of the European Union has how many stars on it?",
      "correct_answer": "12",
      "incorrect_answers": [
        "10",
        "14",
        "16"
      ]
    }
  ]
}



describe(Quiz, () => {
    let playerInfo ={
        userName: "tom",
        difficulty: "easy",
        categoryName: "Animals"
    }

    it('mounts to DOM', () => {
        const div = document.createElement('div');
        ReactDOM.render(<Quiz />, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it('renders correctly', () => {
        const tree = renderer
            .create(<Quiz />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    let wrapper;
    beforeEach(() => wrapper = shallow(<Quiz />));
    it('Renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
    })



    
    beforeEach(() => wrapper.setState({quizInstance: testData.results}))
    it('should render 5 <Question /> if quizInstance state is not null',()=>{

        //let wrapper = mount(<Quiz />)
      
        console.log(playerInfo)
        wrapper.setState({quizInstance: testData.results})
        wrapper.setProps({playerInfo:{
            userName: "tom",
            difficulty: "easy",
            categoryName: "Animals"
        }})
        
        expect(wrapper.find(Question)).toHaveLength(5)
    })

    it('should render a <div />', () => {
        expect(wrapper.find('div').length).toEqual(3);
    });



})



