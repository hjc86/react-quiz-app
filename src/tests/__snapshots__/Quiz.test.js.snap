// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz Renders correctly 1`] = `ShallowWrapper {}`;

exports[`Quiz renders correctly 1`] = `"waiting for 0s round to load load"`;
